streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java

```diff
@@ -447,11 +453,106 @@ public String getStoreForChangelogTopic(final String topicName) {
+                log.error("{}Consumer's subscription does not match the known source topics.\n" +
+                        "consumer subscription: {}\n" +
+                        "topics in topology metadata: {}",
+                    logPrefix, topics, sourceTopicCollection());
+                throw new IllegalStateException("Consumer subscribed topics and topology's known topics do not match.");
+            }
+        }
     }
 
-    void addSubscribedTopicsFromAssignment(final List<TopicPartition> partitions, final String logPrefix) {
-        applyToEachBuilder(b -> b.addSubscribedTopicsFromAssignment(partitions, logPrefix));
+    void addSubscribedTopicsFromAssignment(final Map<TaskId, Set<TopicPartition>> tasks, final String logPrefix) {
+        final Map<String, Set<String>> assignedTopicsByTopology = new HashMap<>();
+
+        for (final Map.Entry<TaskId, Set<TopicPartition>> task : tasks.entrySet()) {
+            final String topologyName = getTopologyNameOrElseUnnamed(task.getKey().topologyName());
+            assignedTopicsByTopology
+                .computeIfAbsent(topologyName, t -> new HashSet<>())
+                .addAll(task.getValue().stream().map(TopicPartition::topic).collect(Collectors.toSet()));
+        }
+        for (final Map.Entry<String, Set<String>> assignedTopics : assignedTopicsByTopology.entrySet()) {
+            final Set<String> newTopics =
+                lookupBuilderForNamedTopology(assignedTopics.getKey())
+                    .addSubscribedTopicsFromAssignment(assignedTopics.getValue(), logPrefix);
+            assignedTopics.getValue().retainAll(newTopics);
+        }
+        updateAndVerifyNewInputTopics(assignedTopicsByTopology);
+    }
+
+    private String getTopologyNameOrElseUnnamed(final String topologyName) {
+        return topologyName == null ? UNNAMED_TOPOLOGY : topologyName;
+    }
+
+    private void updateAndVerifyNewInputTopics(final Map<String, Set<String>> newTopicsByTopology) {
+        final Set<String> duplicateInputTopics = new HashSet<>();
+
+        for (final Map.Entry<String, Set<String>> newTopics : newTopicsByTopology.entrySet()) {
+            final String topologyName = newTopics.getKey();
+            for (final String newTopic : newTopics.getValue()) {
+                if (allInputTopics.contains(newTopic)) {
+                    duplicateInputTopics.add(newTopic);
+                    final String namedTopologyErrorMessagePrefix = UNNAMED_TOPOLOGY.equals(topologyName) ?
+                        "" :
+                        "Topology {}: ";
+                    log.error("{}Cannot add topic {} to the subscription as the application is " +
+                        "already consuming from this topic elsewhere in the topology",
+                        namedTopologyErrorMessagePrefix, newTopic);
+                } else {
+                    allInputTopics.add(newTopic);
+                }
+            }
+        }
+        if (!duplicateInputTopics.isEmpty()) {
```

Comment:
This is unrelated to this PR but the check brings up a good point. Maybe instead of throwing an error on the pattern subscription finding an overlapping topic and then restarting the thread and trying again. Maybe we can pause a topology until the topics are not longer overlapping?

EDIT: I can see this in the other PR